### ISSUE TITLE
Plugin rule invoked twice

### ISSUE THREAD

**Issue by flaviotordini** [markdown-it/markdown-it#884]
I'm writing a plugin that renders YAML blocks using some templating mechanism (like moustache). I've got the rule working, despite the lack of documentation... but my rule is being called twice and I'd like to fix it. Here's my code:

```
const plugin = (md, options) => {

    function getLine(state, line) {
        const pos = state.bMarks[line];
        const max = state.eMarks[line];
        return state.src.substring(pos, max);
    }

    const replaceYAML = (state, start, end, silent) => {
        if (state.blkIndent !== 0 || state.tShift[start] < 0) {
            return false;
        }
        if (getLine(state, start) !== '---') {
            return false;
        }

        let cursor = start + 1;
        const dataStart = cursor;
        let dataEnd = dataStart;
        while (cursor < end) {
            if (state.tShift[cursor] < 0) {
                break;
            }
            if (getLine(state, cursor) === '---') {
                break;
            }
            dataEnd = cursor;
            cursor++;
        }

        const dataStartPos = state.bMarks[dataStart];
        const dataEndPos = state.eMarks[dataEnd];
        const src = state.src;
        const yaml = src.substring(dataStartPos, dataEndPos);
        console.log('yaml', yaml);

        const templateDir = 'TODO';
        const templatePath = 'TODO';
        const templateOutput = 'TODO';

        const startPos = state.bMarks[start];
        const endPos = state.eMarks[cursor];
        state.src = src.slice(0, startPos) + templateOutput + src.slice(endPos, src.length);

        return true;
    };

    md.block.ruler.before('code', 'object', replaceYAML, { alt: [] });
}

module.exports = plugin;
```

The input Markdown is something like:

```
# Title

---
myyaml: stuff
---

Some more markdown
```

Why is the rule is invoked twice?

**Comment by puzrin** [comment]
The most close rule to your case can be "fenced block" (start_line + non_markdown_data + end_line). I'd suggest to take it as base and drop all unneeded parts.

Take care of `silent` param https://github.com/markdown-it/markdown-it/blob/master/lib/rules_block/fence.js#L41. It's missed in your code.

Each rule is called twice:

1. "Silent" - to check if it worth to continue with more deep parse. That's required for performance.
2. "Full" - when content processing really happens.

**Comment by flaviotordini** [comment]
Thanks. It get called twice with silent=false.

I'm having a hard time because of lack of documentation about rule functions and the state object. Is there any documentation I am missing?

**Comment by puzrin** [comment]
It's very difficult to create doc about this kind of parser, because some points are very cryptic. For advanced features reading of source code is the best way to understand what happens. Also, feel free ask questions here, i will try to help.

**Comment by flaviotordini** [comment]
Thanks Vitaly. I think what I'm doing could be a useful plugin for markdown-it users. So I will ask more questions here after I studied a bit more.

**Comment by flaviotordini** [comment]
Hi Vitaly, I endend up writing this. https://github.com/flaviotordini/markdown-it-yaml

I'm planning to extend it a bit in the next days and publish it to npm.

**Comment by rlidwka** [comment]
TL;DR: instead of

```js
state.src = src.slice(0, startPos) + templateOutput + src.slice(endPos, src.length)
```

should be something like:

```js
state.tokens.push(...state.tokens.concat(state.md.parse(templateOutput))) // or something similar
state.line = cursor
```

**Comment by flaviotordini** [comment]
Thanks @rlidwka, in the meantime I understood this better, fixed my code and got everything working.
